### ISSUE TITLE
Tunnel Navigation has jumping GPS location

### ISSUE THREAD

**Issue by Mattsela** [osmandapp/OsmAnd#20485]
### Description

I noticed following behaviours in car navigation mode in tunnels. 

- when navigating and going into tunnel, GPS is lost

Now, quite often, immediately after a tunnel, position jumps to somewhere 5200km far away position (Svalbard) for a second and returns immediately to correct position (in CH, in my case). However, navigation stops as there are not maps downloaded between current position and Svalbard. It doesn't create the route from current position to the goal, but asks to download the missing maps between Svalbard and current position (that would be a lot of maps) . There is no way to recover the navigation w/o downloading. So the navigation needs to be cancelled and started a new navigation event from scratch. This is very annoying when there are many tunnels one after another. This happens with the current version (4.8.2), I didn't observe that earlier (4.7.2 or earlier, however, I reinstalled the app as there were also other issues earlier) . I've tried
- starting the app, device
- reinstalling maps
- reinstalling app
- various settings : location source from play/android api, kalman filter, Map rendering. No impact

For me it looks like there is some kind of default position (Svalbard) where position jumps just when it is about to recover GPS data. Just losing it doesn't cause this jump as this happens always when just coming out of a tunnel. 

When position had jumped for s short moment to Svalbard, app requests maps between Svalbard and current position (or target) and this request can't be just cancelled/ignored but navigation has to be cancelled (or maps downloaded). Easy solution could be a possibility to ignore the download request and continue navigation from current position until a more robust solution is found. Anyway, that option would be handy

### Steps to reproduce

See description, driving through a tunnel

### Actual result

See description, navigation event needs to be restarted 

### Expected result

Continue navigation from current position

### Your Environment (required)

```
OsmAnd Version: 4.7.2 - 4.8.2
Android/iOS version: 12

**Comment by yuriiurshuliak** [comment]
Related to https://github.com/osmandapp/OsmAnd/issues/18408 https://github.com/osmandapp/OsmAnd/issues/19728

**Comment by Guylby** [comment]
I had the same bug in previous versions but not as severe:
When exiting the tunnel, I'd say 1 every 3 times the gps would reset in the middle of the ocean with Osmand saying "You deviated from your route 5500kms". A reroute got triggered, but after 1-2s the gps got back to the true position and canceled the reroute while normal routing resumed.
It never canceled routing asking to download maps though (maybe because my default gps is in the middle of the ocean so no maps exists to be downloaded), but I'll look out for it next time to see if it also does this on my device

**Comment by yuriiurshuliak** [comment]
To help us check and identify this problem, we need a GPX track recorded using the trip recording feature, when the bug occurred on this route. Additionally, try our beta version and let us know if the issue persists: [OsmAnd Beta Versions](https://osmand.net/docs/versions/nightly_versions).

**Comment by Guylby** [comment]
I haven't been able to reproduce the issue with simulating a gps loss during navigation without moving (ie going to the basement so that gps is lost and going back up). I'm guessing you have to move a fair bit (a few hundred meters at least) without a gps fix for the issue to pop up.

On my end, it seemed like as I exit the tunnels (and I drifted far enough away that the previous gps position is not correct anymore) and the gps starts to get signal back from the satellites, there is a small moment when it's not yet synchronized and it indicates a widly erroneous position (for instance, a point 5500kms away in the middle of the ocean) and this gets passed to osmand that produce the "teleportation". Then gps is synchronized and everything gets back to normal. 

I mainly use the gps navigation without any mobile data, so I don't have A-GPS available to get a quicker fix of the position. Maybe a quicker fix from A-GPS hides this issue for most users?

**Comment by Mattsela** [comment]
That is probably how it works. A-GPS likely helps. I have a car unit, which is using a WiFi connection, but in some cases the mobile network is lost in tunnels and then an erroneous position is indicated for a second. When I didn't have WiFi connection, problem occurred basically every time.

**Comment by vmicho** [comment]
I have same issues. Definitely in 4.8.x, maybe also in later 4.7 - i am for a second "you have been off the route for 5301 km". Sometimes (more often in big tunnels) after getting out of tunnels.

I have a screen recording (nothing to be seen, just a 1 second pause in map moving). Sometimes it points very breefly to a place outside of the world (not on my recording and somehow the Screen Recorder does not record audio).

I also have gpx, there I see also nothing:
[2024-08-22_15-12_Thu-bug.gpx.txt](https://github.com/user-attachments/files/16735531/2024-08-22_15-12_Thu-bug.gpx.txt)

From the gpx it looks like GPS position is correct.

I can try to record few more routes next time.

**Comment by Rowin63** [comment]
I will not continue in #18408 (closed) but here.
My screen video is in this other issue and it´s still the same behaviour.
Here´s the recorded track from passing this tunnel.
This 2km Tunnel in Vienna has several exits and if you don´t know where to go by yourself, you´re lost.
Osmand can´t tell you because it still goes nuts without gps signal.

[2024-08-27_15-36_Tue.gpx.txt](https://github.com/user-attachments/files/16763954/2024-08-27_15-36_Tue.gpx.txt)

**Comment by Mattsela** [comment]
Here is a gpx file. The glitch happened in the 3rd tunnel and in the last
[2024-08-28_08-55_Wed.gpx.Txt](https://github.com/user-attachments/files/16780545/2024-08-28_08-55_Wed.gpx.Txt)

**Comment by yuriiurshuliak** [comment]
The bug was not reproduced.  I followed these steps using the provided track "2024-08-28_08-55_Wed.gpx":

- Disabled GPS and enabled flight mode on the device.
- Built navigation using the specified track.
- Activated navigation simulation.
- Waited until entering the tunnel.

As expected, the location simulation worked correctly in the tunnel despite the lack of a GPS signal. 

However, the second video I attached highlights an issue where the track appears to have been recorded as a straight line, bypassing the tunnel without simulating the location.

OsmAnd~ 4.9.0#3096m, released: 2024-08-29


<table style="width:100%">
  </tr>
  <tr>
    <td> <video source src="https://github.com/user-attachments/assets/31d467b9-67f5-456d-8f78-57957f5b4ec5"  width="300" height="auto"/>   </td>
    <td><video source src="https://github.com/user-attachments/assets/c23e5614-d8de-46a1-af8a-27fb7a3e36e0"  width="300" height="auto"/>   </td>
  </tr>
</table>

**Comment by yuriiurshuliak** [comment]
The bug was not reproduced. I tested this on several recorded tracks that were attached above, and in most cases, tunnel navigation worked as expected. However, in a few instances, specifically when entering the tunnel, the navigation stopped, and a voice message announced "GPS Signal lost" (as seen in the fourth video).

That said, the navigation resumed once exiting the tunnel and continued to guide along the correct route.

In conclusion, regarding the original issue of GPS location jumping several kilometers, I did not encounter this problem. However, there is an issue with GPS signal loss and the simulated mapping of navigation through tunnels.

OsmAnd~ 4.9.0#3193m, released: 2024-09-16

<table style="width:100%">
  </tr>
  <tr>
    <td> <video source src="https://github.com/user-attachments/assets/5ea5bdc5-c6cb-4a75-a54b-3d7980816e70"  width="300" height="auto"/>   </td>
    <td><video source src="https://github.com/user-attachments/assets/85d9737c-da44-4b0c-abd4-eca40f807190"  width="300" height="auto"/>   </td>
    <td><video source src="https://github.com/user-attachments/assets/acc478dd-f82b-4144-ad3b-cafe2417945f"  width="300" height="auto"/>   </td>
    <td><video source src="https://github.com/user-attachments/assets/858c4028-9025-4fa5-8ced-782346551bb5"  width="300" height="auto"/>   </td>
  </tr>
</table>

**Comment by Guylby** [comment]
Hi!

I have several new experiences with this bug:

First it seems related to the "Tunnel warning" signal somehow, because the bug will only show if Osmand shows the Tunnel warning sign on the bottom left. Maybe it is related to the tunnel navigation logic if gps signal is lost? (issue #18408)

Let me explain: one time I had routing on a certain route, but changed mind and took a really long tunnel. Osmand did not have time to recalculate and "thought" that we were going through a normal road. Gps signal was lost and when I got out of the tunnel, the bug didn't show up and navigation resumed on the new position. However, when going through below an overpass a few kilometers after (so a really small tunnel with no gps lost), Osmand showed the "Tunnel warning" sign and vocal instruction, but just after getting out of this overpass (which was like 3m long), the bug happened! Osmand told "You have deviated from your itinerary for 5560kms" when the gps position was really accurate on the screen and it did not lost signal!

Second experience on a recent trip in Norway:
Got on a road trip through the fjords, and took approximately 40 differents tunnels. The bug appeared on approximatly 1/3 of the tunnels (so about 15 times) which was really annoying. 
However the deviation was not the same as on my normal area, here it was always "you have deviated from your itinerary for 3400kms"
1 time it did the same as on the first example: got through the tunnel fine, but got below a small overpass (with no gps signal loss) but there was the osmand "Tunnel warning" sign and after crossing it, the bug happened with vocal instruction saying "you have deviated from your itinerary for 3400kms"

I can also say that in most tunnels the simulated position when the gps signal is lost did not work like in issues #18408 and #19728

3 times, the navigation stopped because the imaginary point 3400kms away was on an area that I did not have maps for, so routing just stopped asking to download missing maps. I had to manually reset routing which was also annoying. 

All of this was on Osmand google beta 4.9.0

**Comment by mwerle** [comment]
Currently : OsmAnd 4.8.6, Android 13, not using A-GPS, just GPS (phone does not have wifi enabled, nor a sim card - I use this as a completely off-line mapping solution).

Have been getting this issue for nearly a year in Japan.

Some points from a user perspective:

1. GPS jumping to a far away location should be ignored. It is -patently- incorrect if the new location is 5000+kms away in only a second (heck, even a minute or an hour). I thought there was a fix for this issue a while back, but it seems it continues cropping up.
2. Routing display should NEVER break - the current route should remain until a new route has been successfully calculated in the background and then seamlessly swapped. Possibly have some sort of indicator that a recalculation is occurring, but don't delete the old route first. Having the old route displayed, even if no updated position information is available, at least allows the user to try to navigate manually.
3. "Oboslete" maps should never prevent routing - possibly pop up a warning or make it a user-option, but simply stopping a route and complaining about old maps is extremely poor user experience - old maps and possibly incorrect routes are better than no routes at all.

Background: I use OsmAnd exclusively as my motorcycle navigator. When it works, it's great, however recently it has become more and more flakey and occasionally has put me in quite a predicament (Tokyo inner city expressways, for example, where it is practically impossible to stop to reset the navigation and there are many tunnels). 

----

Here's a GPX file during which I had this issue a couple of times. Interestingly the GPX file does not show the issue at all.

At the beginning of the track the GPS was almost completely broken; no idea why. From 11:12:39 to around 12:10:36 the GPS track is all over the place, often ending up in the ocean or quite far from the roads I was actually travelling on. Once it had "come good" it was fine, except tunnels.  There were a lot of tunnels between ~14:16:02 to 14:37:47; at least one of which resulted in complete loss of routing. The track file does not appear to reflect this.

[Mon 21 Oct 2024.gpx.zip](https://github.com/user-attachments/files/17463110/Mon.21.Oct.2024.gpx.zip)

**Comment by scaidermern** [comment]
I just had the same problem while driving through a tunnel with OsmAnd 4.9.0. Somewhere in the middle of the tunnel OsmAnd suddenly decided that I'm no longer on the correct track and started to recalculate the route (which never finished).

Location source is set to Android API, so this can't be caused by a faulty location from network sources.

This was the first time in years I had this problem. Maybe something broke in recent OsmAnd versions. Or it was just a happy little accident.

**Comment by Rowin63** [comment]
The problem in a tunnel has changed, now this happens (I´m inside the tunnel, following the "red arrow" lane):

![Screenshot_20241022_094841_OsmAnd+](https://github.com/user-attachments/assets/56203f1b-b48f-4722-b42c-6d2b5b41ecbb)

**Comment by mwerle** [comment]
> In conclusion, regarding the original issue of GPS location jumping several kilometers, I did not encounter this problem.

Could this possibly be due to a particular implementation of Android and/or physical device which may have different ways (or contain bugs) when dealing with lost GPS signal? 

A way to simulate this would be to manually insert data into a GPX file with locations thousands of kilometers away, and then seeing how OsmAnd deals with it. My prediction is that OsmAnd stops routing due to "missing maps" and then refuses to recalculate after the location returns to the true location.

**Comment by Guylby** [comment]
Maybe Osmand could have a check on gps loss when it regains signal to check the delta between the new gps ping and the previous ones, and if the delta is too big (like >1000kms), just ignore it? Just to ignore outliers until signal is stable again?

**Comment by Guylby** [comment]
Today I made a detour to take a long tunnel to try to catch the bug and it happened on the first try!
I recorder a .gpx but unfortunalty nothing shows on it, I used the smallest recording interval (Continuous), but it still logged only every second or so the position, so there was no huge deviation, but I have the screen recording showing the position jumping >5500kms away:
For some reason sound wasn't recorded but it said "You have deviated from your position for 5760kms". 
This is just the end of the tunnel, gps signal was 0/16 for the duration of the tunnel, but the simulated position worked quite well as it was really close to the true position when I exited and regained signal. 


https://github.com/user-attachments/assets/582173ae-46da-4f65-92ec-5965d840f3e2


Here is the gpx track with just the .zip extension added, but the deviation doesn't show:
[2024-10-29_13-57_Tue.gpx.zip](https://github.com/user-attachments/files/17560344/2024-10-29_13-57_Tue.gpx.zip)

**Comment by Guylby** [comment]
I just noticed that I had <50m as the minimal gps precision in the track recording settings so maybe the déviation wasn't recorded because of it

**Comment by scaidermern** [comment]
> Maybe Osmand could have a check on gps loss when it regains signal to check the delta between the new gps ping and the previous ones, and if the delta is too big (like >1000kms), just ignore it? Just to ignore outliers until signal is stable again?

A sanity check for locations has already been suggested in various tickets but so far it never got implemented.

**Comment by scaidermern** [comment]
> we didn't do any cleaning for jumping locations cause you need to choose one which is correct A or B.

Maybe a moving average can help. This will filter out single outliers. It will also allow big jumps after a slight delay if the next `n` locations are equally far away from the last valid one.

**Comment by vshcherb** [comment]
Bug has been reproduced and indeed there is no location written inside GPX file which raises suspicion that location somehow get from FusedLocationProvider or NetworkLocationProvider and needs to be filtered certainly in navigation especially in the tunnel

**Comment by vmicho** [comment]
Hi.
I've managed to get a screenshot. Not the coords as my coords widget is off during navigation and could not get coords from the position (blue dot) as the position was not known (I only got a list of all my markers and favourites after clicking anywhere)
Coords look something like -90N 180E (comparing from OSM)

![Screenshot_20241109-151811_OsmAnd+](https://github.com/user-attachments/assets/a801317f-8a28-46bd-9015-432dd28aaa55)

**Comment by vshcherb** [comment]
For Android Auto there is a potential fix for incorrect double location - https://github.com/osmandapp/OsmAnd/commit/7bf1b22a51563eb091a24cbc8da768400f65e57f.

**Comment by Zirochkabila** [comment]
`OsmAnd~ 5.0.0#3764m, Released: 2024-11-19`

49°47'23.4"N, 18°06'40.2"E

Tested with the latest version of OsmAnd, but the behavior is not reproduced. 
Seems to have been fixed

**Comment by Zirochkabila** [comment]
Bug is still present

**Comment by mwerle** [comment]
> Bug is still present

Yup; I got screenshots today, OsmAnd jumped 5000+km while I was briefly stuck in an underpass..  

Using the OsmAnd beta - 4.9.7 2024-11-18

![Screenshot_20241123-132544](https://github.com/user-attachments/assets/25fb4384-1f0c-4c22-967f-8417bff43ad7)
![Screenshot_20241123-132859](https://github.com/user-attachments/assets/09e9b310-bb78-4431-8f37-ff0827fdd53c)

**Comment by Mattsela** [comment]
I've tested it with

4.9.2: in about 40 tunnels it once jumped 5000km, but navigation didn't
break and it continued as it should be
4.9.4: about 20 tunnels without problems
4.9.7: about 10 tunnels without problems

Tunnel simulation doesn't work (never has) but position is somewhere close
to it. When coming out of a tunnel and again receiving GPS, navigation
continues (now with 4.9.2 or newer)


Regards
Matti

**Comment by Rowin63** [comment]
Different behaviour in Nightly/Android release, after losing satellite signal:

Playstore: jumps to a position >5000km away
Nightly: jumps to a nearby road and restarts route calculation

Both start correct recalculation as signal is back

**Comment by Mattsela** [comment]
Further test with 4.9.7.

In about 10 tunnels it jumped to 5000km far away in one tunnel twice but
continued navigation correctly after receiving GPS signal



Matti

**Comment by vshcherb** [comment]
Today was testing again, there was a sudden appearance that route is far away but then suddenly it was back to good simulation so none recalculation happened and none voice announcement. Though it was visible dashed turn type on the screen.

**Comment by Zirochkabila** [comment]
OsmAnd~ 5.0.0#3780m, Released: 2024-11-21
Android 14

[OsmAnd](https://osmand.net/map/search/#17/48.36061/16.38165)
[OSM](https://www.openstreetmap.org/search?query=48.36061%C2%B0%20N%2C%2016.38229%C2%B0%20E#map=17/48.360819/16.383083)

[2024-11-22_11-13_Fri.gpx.zip](https://github.com/user-attachments/files/17901628/2024-11-22_11-13_Fri.gpx.zip) Track that was recorded during the trip


`48.36061° N, 16.38229° E` place shown on the video

https://github.com/user-attachments/assets/3f9e870c-b73c-4b87-b875-600f90e001c6

**Comment by Rowin63** [comment]
This "new variant" mentioned above is reproduceable. Again I had to stop and restart the route, because after the tunnel the routing line disappeared.

**Comment by vshcherb** [comment]
Filtering locations is introduced however the issue might not yet fixed

**Comment by vshcherb** [comment]
To further investigate situations, please record track but change default setting of Accuracy to 0! (Not selected) - in that case we could find what accuracy points are triggered (or not triggered and causing issues)

**Comment by vshcherb** [comment]
Strangely enough: trip recording was not taking location from service (when phone screen is off) - https://github.com/osmandapp/OsmAnd/commit/7046d9c25f3ca6ed322bc153961f08f3f9dd84ee. I've added it, it shouldn't cause duplicate locations but it's worth to test it.
CC @sonora

**Comment by mwerle** [comment]
@vshcherb If you're not convinced this issue is fixed, why did you close it? Perhaps it would be good to give affected people time to confirm the fix first. Thank you. (FWIW, I won't be able to confirm this fix myself for a while as my bike is now in winter storage..)

**Comment by vshcherb** [comment]
Cause I need to bring attention to retest it and provide more details, everything of code inspection and fixing small potential issues is done

**Comment by Zirochkabila** [comment]
OsmAnd~ 5.0.0#3853m, Released: 2024-11-30    --------              **Bug still present** The behavior is still repeated. If the tunnel is long, the route is recalculated, if the tunnel is short (as in the photo), the route is not recalculated. But, in both cases, the GPS signal is lost, and the location jumps to another place

[2024-12-01_15-16_Sun GPS ERROR POSITION TUNEL.gpx.zip](https://github.com/user-attachments/files/17975828/2024-12-01_15-16_Sun.GPS.ERROR.POSITION.TUNEL.gpx.zip)

[2024-12-01_15-28_Sungпрлдзжжло.gpx.zip](https://github.com/user-attachments/files/17975721/2024-12-01_15-28_Sung.gpx.zip)



<table style="width:100%">
  <tr>
    <th>48.14336° N, 13.98971° E</th>
    <th>48.27124° N, 14.29574° E</th>
  </tr>
  <tr>
    <td><video source src="https://github.com/user-attachments/assets/b86f1e18-cfaf-4cc5-b305-014f1edfc127"  width="400" height="auto"/></td>
<th><img src="https://github.com/user-attachments/assets/99e918b2-65a8-4707-af0c-ffd54f72a6f6"  width="300" height="auto"/> </th>
  </tr>
</table>

**Comment by vshcherb** [comment]
Adroid APi was fixed - separately (passive provider is no longer used)
Google Locations fixed by filtering noisy fused locations from provider

**Comment by sonora** [comment]
> Android APi was fixed - separately (passive provider is no longer used)

Any potential effect / risk on battery consumption we need to watch out for in testing?

**Comment by Zirochkabila** [comment]
# Testing 5


It was tested on two versions of OsmAnd using Google services and API. It seems that Google services have started to work as an API (in the previous testing) - when using both types of resources, the location is saved, the route is not recalculated, the route does not disappear.

`OsmAnd OsmAnd 4.9.9, Released: 2024-12-06 `

[Google (1).gpx.zip](https://github.com/user-attachments/files/18063857/Google.1.gpx.zip)
[API (2).gpx.zip](https://github.com/user-attachments/files/18063859/API.2.gpx.zip)


<table style="width:100%">
  <tr>
    <th>Google</th>
    <th>API</th>
  </tr>
  <tr>
    <td><video source src="https://github.com/user-attachments/assets/90c5aaf3-9bd4-4a55-8fae-754e0eaef47c"  width="300" height="auto"/></td>
    <td><video source src="https://github.com/user-attachments/assets/df04a779-ddf9-416a-89c4-43610eda4ad8"  width="300" height="auto"/> </td>
  </tr>
</table> 


`OsmAnd~ 5.0.0#3902m, Released: 2024-12-07`

[Google (2).gpx.zip](https://github.com/user-attachments/files/18063933/Google.2.gpx.zip)
[API (3).gpx.zip](https://github.com/user-attachments/files/18063934/API.3.gpx.zip)

<table style="width:100%">
  <tr>
    <th>Google</th>
    <th>Api</th>
  </tr>
  <tr>
    <td><video source src="https://github.com/user-attachments/assets/9570b122-3577-407b-8319-53c64088f874"  width="300" height="auto"/></td>
    <td><video source src="https://github.com/user-attachments/assets/c40c0c23-d2e5-4f41-8b2e-ef5c73ca7e64"  width="300" height="auto"/> </td>
  </tr>
</table> 


However (**BUGS**)

1. After the car has entered the tunnel for 50 meters (+-), the track stops displaying on the map (perhaps the track stops being recorded as in the location simulation) and starts displaying only after the car leaves the tunnel (the track recording process is visible in all videos)
2. Also, if you reopen OsmAnd (open the camera and return to navigation), the location stops updating (freezes at the point where you switched to the camera) and continues to update after leaving the tunnel (freezing on the video below)


https://github.com/user-attachments/assets/dc888f9e-f773-4419-9552-d08b76645d9f

**Comment by Zirochkabila** [comment]
@Rowin63 Did you record gpx track?
It would be nice to have it with Google resources & API (for comparison)

**Comment by Rowin63** [comment]
Sorry, no.
I'll try to keep it in mind and do a recording next time I'm around 😉

**Comment by Zirochkabila** [comment]
@Rowin63 You can use this profile to check it - it uses the data I tested with
[Driving Zirochka tets.osf.zip](https://github.com/user-attachments/files/18113089/Driving.Zirochka.tets.osf.zip)

**Comment by Rowin63** [comment]
> @Rowin63 You can use this profile to check it - it uses the data I tested with [Driving Zirochka tets.osf.zip](https://github.com/user-attachments/files/18113089/Driving.Zirochka.tets.osf.zip)

I would like to do so - but I don´t get the profile imported...
How to do it? ("settings > import from file" does not work...?)

**Comment by Zirochkabila** [comment]
@Rowin63 To begin with, unzip and then open it from a file manager (or we have a new bug with importing) or send the unzipped file to a messenger (Telegram, for example) and import from there

**Comment by Rowin63** [comment]
> @Rowin63 To begin with, unzip and then open it from a file manager (or we have a new bug with importing) or send the unzipped file to a messenger (Telegram, for example) and import from there

Aaah, my (stupid) mistake. Didn´t unzip it, thought Osmand would take it as it is...

**Comment by Rowin63** [comment]
> @Rowin63 You can use this profile to check it - it uses the data I tested with [Driving Zirochka tets.osf.zip](https://github.com/user-attachments/files/18113089/Driving.Zirochka.tets.osf.zip)

@Zirochkabila Using your profile:

) two short tunnels, in each the position jumped away 5200km (track "Stadlau")

![Screenshot_20241213_085722_OsmAnd+](https://github.com/user-attachments/assets/72bb168e-c3b9-4c68-9b1f-4ef8591c171d)

![Screenshot_20241213_085813_OsmAnd+](https://github.com/user-attachments/assets/467a6018-afbd-4234-87d8-88b2dcd76b30)

[2024-12-13_Stadlau.gpx.txt](https://github.com/user-attachments/files/18127468/2024-12-13_Stadlau.gpx.txt)


) one long tunnel, position was always nearly accurate on display, only a bit jerky (track "Kaisermühlen")
I´m a bit surprised about this result (keeping position)  because this tunnel is more than 2km long

[2024-12-13_Kaisermühlen.gpx.txt](https://github.com/user-attachments/files/18127471/2024-12-13_Kaisermuhlen.gpx.txt)

(added .txt for uploading)

**Comment by Zirochkabila** [comment]
@Rowin63 I just checked in two different tunnels (with different lengths), on two different versions (release and nightly), using Google & API - your result, at least for now, is not repeatable. Everything works as in this commentary https://github.com/osmandapp/OsmAnd/issues/20485#issuecomment-2528449850
We will continue to monitor the development of this issue.

**Comment by vshcherb** [comment]
@Rowin63 probably you didn't enable minimum accuracy - not specified, so your recorded track doesn't contain information about gps provider and offset location. Or it's possible you didn't use latest version with fixes and recording informaiton.

**Comment by Rowin63** [comment]
Yesterday I ran the +4.9.9

**Comment by Rowin63** [comment]
What settings should I use?

**Comment by vshcherb** [comment]
Record settings - minimum accuracy - not specified (0). Also probably have development plugin on

**Comment by Zirochkabila** [comment]
@vshcherb All these settings were set in the profile provided for testing 
The track provided by @Rowin63  may not contain information because the track continues recording only after leaving the tunnel

**Comment by Zirochkabila** [comment]
`OsmAnd~ 5.0.0#3972m, Released: 2024-12-20` - An additional test run was conducted (perhaps the behavior with Overlay could have had an impact, as mentioned here https://github.com/osmandapp/OsmAnd/issues/21453#issuecomment-2556438042) - to no avail. Navigation in the tunnel works.

**Comment by scaidermern** [comment]
I just experienced this problem today in one (out of several) tunnels. Suddenly OsmAnd thought I was 4800 km away.

OsmAnd 4.9.10, Android 13, Fairphone 4.

**Comment by Mattsela** [comment]
I had also today the 5200km jumping issue in one tunnel.
Navigation didn't crash but continued working OK. Also tunnel simulation
now works quite well.
Version 4.9.10



Matti

**Comment by Mattsela** [comment]
Now on two trips Navigation crashed, no possibility to recover.
Error message the usual one, missing maps.
Version 4.9.10


Matti

**Comment by Guylby** [comment]
On 4.9.10,

The "You have deviated 5500kms..." still happens as often for me (with Google Api)

On 20% of these occurences navigation is stuck, with the popup "Download missing maps" blocking everything. You can't cancel it or ignore it you have to manually stop routing, and start it again to get back to navigating

**Comment by Mattsela** [comment]
For me this happens still at least once per trip but not in all tunnels.
The basic problem still exists
Version 4.9.10


Matti

**Comment by Rowin63** [comment]
Tunnel navigation still does not work reliable. 
It still is not possible to follow a navigation with exits in a tunnel - you get nothimg but lost.

Two tunnels, two different happenings today.

Tunnel 1: 
Position was lost and at the end of the tunnel I was placed 5000km away.
After some moments position was found and routing continued correctly.
=> if this happens, it seems to happen always at the end of a tunnel, just like when getting gps signal again!

![Image](https://github.com/user-attachments/assets/098a8030-805d-4303-85ba-16afe9f50f4b)

Tunnel 2:
"Effect 1" (jumping 5000km) did not happen, but signal was lost and when getting gps signal again, the routing was continued - but without routing line displayed on the map. To "solve" this, one has to stop navigation and start a new navigation.
No screenshot from this, this "effect" was posted before here.

**Comment by Rowin63** [comment]
As this issue was marked as "solved" and closed almost four months ago (Dec 10th) it should not exist anymore in the +5.0.2 - but it still does.

**Comment by Zirochkabila** [comment]
@Rowin63 It seems that the problem is localized (it occurs only in certain tunnels)
Please record the track while navigating the tunnels with the following settings in mind:
1. Enable the dev plugin
2. Disable track enhancement
3. Logging interval = 0
4. General logging interval = 0
5. Min. displacement = Not selected
6. Min. accuracy = Not selected
7. Min. speed = Not selected
8. Try Google services & Api (Location source)

**Comment by Rowin63** [comment]
Thanks a lot for the detailed feedback. I'll try this asap. 👍

**Comment by Guylby** [comment]
Can confirm that it still happens on my end too on 5.0.2, had the 5000kms away deviation this morning. I'll do the steps above to try to catch it next time with a track recording

**Comment by Rowin63** [comment]
> @Rowin63 It seems that the problem is localized (it occurs only in certain tunnels)
> Please record the track while navigating the tunnels with the following settings in mind:
> 1. Enable the dev plugin
> 2. Disable track enhancement
> 3. Logging interval = 0
> 4. General logging interval = 0
> 5. Min. displacement = Not selected
> 6. Min. accuracy = Not selected
> 7. Min. speed = Not selected
> 8. Try Google services & Api (Location source)
> 

Sorry, but I don't find 2 and 3 🤔

**Comment by Zirochkabila** [comment]
@Rowin63 
1. Menu > Plugins > OsmAnd development > Enable
2. Everything related to the appearance and other settings (except for those specified in clauses 3-8)

**Comment by mwerle** [comment]
Still getting this as well.

**Comment by sonora** [comment]
This option has always been around (in some flavors of the app), and yes, plesse try that! It uses the GPS system (offline) more natively and not via google play services.  I have good experience with this option.

**Comment by ronnystandtke** [comment]
I am using OsmAnd 5.0.5 on a Fairphone 5 running LineageOS 22.2 (Android 15) and I think I have found a combination where this problem is easily reproduced:

1. Set the routing type in the development settings to "Highway Hierarchies" (unlike the A* variants, this routing type checks for missing or obsolete cards)
2. Drive through tunnels

I recorded a short video demonstrating the issue:
[ OsmAnd, Tunnels and Routing type "Highway Hierarchies"](https://www.youtube.com/watch?v=XThkWpgZr0g)

Hopefully this will help to reproduce, narrow down, and eventually fix the bug.

**Comment by Guylby** [comment]
This has been the exact same behaviour I'm experiencing in the video you're linking to!

**Comment by sonora** [comment]
It's probably quite obvious, but just to state it here again:

For all reported cases it seems the bogus position GPS seemingly 'jumps' to is the 'null island' 0°N, 0°E.

So very likely this behavior is a peculiarity of some devices' GPS implementation after losing/regaining a position fix.

**Comment by Guylby** [comment]
I don't think it's actually the "null island" but another fixed coordinates: "85.05113° N, 180.00000° E" This location seems to be along the "seam" of the map

I don't remember how, but I got the same jumping gps location bug while routing via simulation! I managed to get a screenshot of the point where the location jumped to with gps coordinates enabled:


<img src="https://github.com/user-attachments/assets/a537e524-abc5-4379-87d5-c21f452fc4e9" width="50%">

Now I've put the center of the map on the location of where I lost signal and got the jumping bug while searching for both "null island" and the other coordinates, and done the same for the location where user Ronnystandtke got his bug, both times the deviation seem very close to the location of the "85.05113° N, 180.00000° E" coordinates:

| Location of my jumping gps location | Map centered on user Ronnystandtke jumping gps location |
| ------------- | ------------- |
| ![Image](https://github.com/user-attachments/assets/d12119bc-821e-45aa-b4b4-5cd7e522c452) | ![Image](https://github.com/user-attachments/assets/5a04c98b-3932-4bef-9393-1c1d2018f4cc) |

**Comment by sonora** [comment]
@Guylby  Indeed, well observed,I think you're right!

**Comment by ronnystandtke** [comment]
Here's another attempt using the "A* 2-phase" routing type:
[OsmAnd, Tunnels and Routing type "A* 2-phase"](https://www.youtube.com/shorts/0HR4C6OB48E)

Although the error message about not being able to calculate the route after exiting the tunnel still appears, the route is successfully recalculated. This is likely due to the "A* 2-phase" routing type.

I also enabled the coordinates widget, which shows that the GPS coordinates briefly jump to a location in the Arctic Ocean.

In fact, it is **exactly** the same position that @Guylby noticed.

**Comment by ronnystandtke** [comment]
There are now many examples showing that this bug is still present in the latest versions of OsmAnd. Could an administrator please reopen the bug report?

**Comment by Guylby** [comment]
I'm also thinking that this bug is not produced by an imprecise gps location from the provider (Google or Android API) but by some internal Osmand logic, 
I tried but was not able to reproduce it, but the instance where the bug appeared and where I was able to get the gps coordinates were not during actual navigation but by using the "Simulate navigation" in the Navigation menu with the phone gps turned OFF. So it was Osmand "providing" the gps signal. 
I also think that this bug is related to the tunnel navigation simulation as the bug only occurs when Osmand knows a Tunnel is coming (ie the tunnel warning appears on screen if enabled), but can also occur on really short tunnels (ie road underpass) where the gps signal isn't even lost. 
My guess would be that the problem occurs when Osmand switches from the internal tunnel navigation gps simulation back to the real gps signal

**Comment by sonora** [comment]
I substantiate the thought that this is probably an Osmand-caused reaction to some GPS-chipset event, but not chipset-created directly.

Because 85.05113°N is the maximum latitude and 180.0000°E is the maximum longitude value when using the Web Mercator projection (also known as EPSG:3857). So likely this comes from the app interpreting something here in this fashion, while I see no reason the GPS chipset should provide exactly these numbers.

**Comment by Mattsela** [comment]
#20485 is not solved. The issue is still there



Matti
